ToolsPanel component: refactor  to typescript

glendaviesnz opened the pull request:
## Description
Converts the ToolsPanel component to typescript. Also adds type annotations to the Menu components that it uses.

## How has this been tested?

- Check that linting and typescript compile checks all complete successfully
- Add margin support to the Group block and add a group block and make sure the dimensions panel all works as expected.

Commit:
Add typing for dropdown menu

Commit:
Add types for ToolsPanelHeader

Commit:
Add remaining types

ramonjd commented on packages/components/src/tools-panel/tools-panel-item/hook.ts:
```diff
@@ -11,8 +11,9 @@ import * as styles from '../styles';
+import type { ToolsPanelItemProps } from '../types';
 
-export function useToolsPanelItem( props ) {
+export function useToolsPanelItem( props: ToolsPanelItemProps ) {
```
Writing a note for myself really, which I hope will be answered during review 😄 

Does this need to hook into [PolymorphicComponentProps](https://github.com/WordPress/gutenberg/blob/2b974ceda8070ee9460f8457dbbd29db4a2abfa8/packages/components/src/ui/context/polymorphic-component.ts)?

E.g., 

```
export function useToolsPanel( props: PolymorphicComponentProps<ToolsPanelProps, 'div'> ) {
```

Just looking at the other hooks and the [recommendations on the original PR](https://github.com/WordPress/gutenberg/pull/32392#discussion_r679139862).

glendaviesnz replied:
Good question, will take a look at that.

glendaviesnz replied:
Have added in the `PolymorphicComponentProps` now.

ciampo commented:
Just an FYI that the `PolymorphicComponent*` types have all been renamed to `WordPressComponent*` in https://github.com/WordPress/gutenberg/pull/34330

A rebase + refactor of all types beginning with `PolymorphicComponent*` should do it!

glendaviesnz commented:
> A rebase + refactor of all types beginning with `PolymorphicComponent*` should do it!

Thanks for letting me know @ciampo

Commit:
Update PolymorphicComponent* references to WordPressComponentProps*

ciampo commented:
Hey @glendaviesnz , for the sake of [coordinating/tracking work](https://github.com/WordPress/gutenberg/issues/34345) related to components and global styles sidebar, do you have an estimate of when this PR is going to be ready for review?

glendaviesnz commented:
> do you have an estimate of when this PR is going to be ready for review?

@ciampo, this was waiting on some refactors of this component that @aaronrobertshaw has on the go - we were planning to take the hit on the conflicts on this PR, rather than complicate those other PRs by merging this first. Aaron, do you have any thoughts on timelines for the other ToolsPanel work that is underway?

ciampo commented:
> @ciampo, this was waiting on some refactors of this component that @aaronrobertshaw has on the go - we were planning to take the hit on the conflicts on this PR, rather than complicate those other PRs by merging this first. Aaron, do you have any thoughts on timelines for the other ToolsPanel work that is underway?

Thank you for the explanation! I guess you're referring to #34530 ?

I understand that you'd rather handle conflicts in this PR than slowing down other open PRs — I'd just like to make sure that this PR doesn't get too de-prioritised, since it looks like the `ToolsPanel` component may keep receiving tweaks in the short term as we iterate on its UX. Converting the component to TypeScript can actually be of help to speed up future refactors and catch early small bugs!

aaronrobertshaw commented:
> Aaron, do you have any thoughts on timelines for the other ToolsPanel work that is underway?

I believe the primary PR is https://github.com/WordPress/gutenberg/pull/34157 with https://github.com/WordPress/gutenberg/pull/34530 secondary to that.

It looks like we [might be able to merge](https://github.com/WordPress/gutenberg/pull/34157#discussion_r703227941) https://github.com/WordPress/gutenberg/pull/34157 and handle passing the tools panel context through the slot's `fillProps` in a separate PR. Either way, that should be very close.

It's less clear how long https://github.com/WordPress/gutenberg/pull/34530 might take. This one is less critical so if needed it could be redone after the TypeScript conversion.

> I'd just like to make sure that this PR doesn't get too de-prioritised

The TypeScript conversion is definitely a priority and the need to make that happen has been discussed a few times in recent days. At this point, the path of least resistance still appears to be as @glendaviesnz suggested, handling the conflicts in this PR and completing the conversion prior to any further tweaks.

For some additional background, the reason for https://github.com/WordPress/gutenberg/pull/34157 being done before the conversion was to attempt to resolve an issue with duplicate "Dimensions" panels for the Cover block in the last release.

ciampo commented:
Thank you for the explanation, @aaronrobertshaw! I'll keep my eyes open for future progress on this PR once #34157 (and follow-ups) get merged!

Commit:
Start refactor of ToolsPanel component to typescript

Commit:
Start typing the ToolsPanelHeader component

Commit:
More conflict resolution

Commit:
Fix type issues following rebase

Commit:
fix remaining menu item type issues

andrewserong commented on packages/components/src/dropdown-menu/types.ts:
```diff
@@ -0,0 +1,66 @@
+	 * human-readable `title` string, `isDisabled` boolean flag and an `onClick` function callback to invoke when the option is selected.
+	 * A valid DropdownMenu must specify one or the other of a `controls` or `children` prop.
+	 */
+	controls?: number;
```
Should this type be an array of objects? (Or add an additional type for the DropdownMenuControl, and make it an array of controls, e.g. `Array< DropdownMenuControl >`?

glendaviesnz replied:
Thanks, I think that was just a bad copy paste on my part - have fixed

andrewserong commented on packages/components/src/dropdown-menu/types.ts:
```diff
@@ -0,0 +1,66 @@
+	 * the same values as given to a [`Dropdown`'s `renderContent`](/packages/components/src/dropdown#rendercontent) (`isOpen`, `onToggle`, `onClose`).
+	 * A valid DropdownMenu must specify one or the other of a `controls` or `children` prop.
+	 */
+	children?: ReactNode;
```
Does this need to be a function type instead that returns a ReactNode? Something like:

```
children?: ( props: any ) => ReactNode 
```

Or also add another type for the props? (Though to be pragmatic, I suppose you might want to leave it as a fairly broad type for the moment to avoid having to continue adding more types in this PR 😅)

glendaviesnz replied:
Another good catch - thanks, fixed.

andrewserong reviewed (commented):
Nice work here @glendaviesnz! I just left a couple of comments about the DropDownMenu props types — I know this is is a WIP, and my TypeScript skills a pretty rusty, so apologies if the comments aren't relevant or if you were in the middle of working on it 🙂

Commit:
Improve menu item types

glendaviesnz commented:
@ciampo this is ready for some review now. This is the first bit of typescript I have done in quite some time, so very open to feedback on this.

ciampo reviewed (commented):
Hey @glendaviesnz , thank you for working on this PR!

I've noticed that this PR includes changes to four components: `DropdownMenu`, `MenuGroup`, `MenuItem` and `ToolsPanel`. I believe we should have **a separate PR for each of these components**, which will the review process a lot easier and more focused:

- I propose we keep this PR focused on `ToolsPanel` — if you get any TypeScript error when importing dependencies into the `ToolsPanel`'s TypeScript files, you can add `// @ts-nocheck` at the start of the imported file to temporarily suppress those warning (see an example of this [for the `Popover` component](https://github.com/WordPress/gutenberg/blob/60d907bc33fcbd47235b8e141909b368966c7995/packages/components/src/popover/index.js#L1))
- We can open follow-up PRs to take care of typing `MenuGroup` and `MenuItem`. Typing the `DropdownMenu` extensively can be complicated, given how many of its props reference other components' props (`popoverProps`, `toggleProps`, `menuProps`) and how the arguments of the `children` prop [depend on the `Dropdown'`s `renderContent` prop](https://github.com/WordPress/gutenberg/blob/81c00480d842383b1e20b39a508540ac27c22854/packages/components/src/dropdown-menu/index.js#L156) — we should skip this for now

---

Apart from that, I noticed that some of the files in the `tools-panel` folder are still with the `.js` extension — ideally, we'd want to convert all JS files to `.ts(x)` (with the only exception of `stories/index.js`).

aaronrobertshaw committed:
Move remaining eligible ToolsPanel files to TS

aaronrobertshaw committed:
Revert "Improve menu item types"

This reverts commit 2676061b2f68577eb65cc977d28baab85694f057.

aaronrobertshaw committed:
Revert "Add typing for dropdown menu"

This reverts commit 81c00480d842383b1e20b39a508540ac27c22854.

aaronrobertshaw committed:
Remove menu group and item type doc block annotations

aaronrobertshaw committed:
Add ts-nocheck to ToolsPanelHeader

The typing for DropdownMenu, MenuGroup and MenuItem will be handled separately and may entail a full TypeScript conversion rather than only doc block annotations.

aaronrobertshaw commented:
> I believe we should have a separate PR for each of these components, which will the review process a lot easier and more focused:

The DropdownMenu, MenuGroup and MenuItem doc block type annotations have been removed from this PR so they can be handled separately as suggested.

> Apart from that, I noticed that some of the files in the tools-panel folder are still with the .js extension — ideally, we'd want to convert all JS files to .ts(x) (with the only exception of stories/index.js).

I've switched the remaining ToolsPanel files from `js` to `ts` (all except `stories/index.js` and `test/index.js`). 

---
After making these changes:
- Unit tests are still passing
- ToolsPanel functions correctly still in storybook and editor
- Prior to adding `ts-nocheck` to the `ToolsPanelHeader` file, the only lint errors were those relating to `DropdownMenu`, `MenuGroup` and `MenuItem` as expected after rolling back those changes.
- After adding `ts-nocheck` there are no remaining linting errors.

@ciampo Would it be best to create new PRs for MenuGroup/MenuItem etc using the doc block annotations removed from this PR as their basis? So that those can be merged quickly, or should the new PRs aim to complete a full conversion to TypeScript for those components?

ciampo reviewed (commented):
> I've switched the remaining ToolsPanel files from `js` to `ts`

Thank you!

> the only lint errors were those relating to DropdownMenu, MenuGroup and MenuItem as expected after rolling back those changes.
> After adding `ts-nocheck` there are no remaining linting errors.

My bad, I haven't explained myself clearly enough. The `// @ts-nockeck` comment shouldn't be added to `ToolsPanel` — it should be added, instead, to all of the dependencies that `ToolsPanel` pulls in (`DropdownMenu`, `MenuGroup`, `MenuItem`, and recursively all of their local dependencies).

These are the steps that I would follow:

1. Remove the `// @ts-nocheck` comment from `ToolsPanelHeader`
2. Add the `src/tools-panel/**/*` folder to [`tsconfig.json`](https://github.com/WordPress/gutenberg/blob/bdb92729488eb33aed0b2ed949e9236153911f06/packages/components/tsconfig.json). **This step is actually necessary to add the component to the TypeScript project**, and thus enable TypeScript checks (so far they were not! My bad for not noticing this until now)
3. Run `npm run dev` and find all the errors that refer to files outside of the `src/tools-panel/` folder. We should add `// @ts-nocheck` at the start of each of those files, and add their folders to `tsconfig.json` (similarly to what done in step 2). In our case, the folders should be `src/dropdown/**/*`, `src/dropdown-menu/**/*`, `src/menu-item/**/*`, `src/menu-group/**/*` and `src/navigable-container/**/*`.
4. At this point, the `MenuGroup`, `MenuItem` and `DropdownMenu` components are still showing errors inside `ToolsPanelHeader`. To solve them, we need to slightly refactor these components and move the `props` destructuring away from the function signature — this is to prevent TypeScript from inferring the types of these props. For the same reason, we will also need to temporarily delete the JSDoc types of `MenuItem` (but we will add those back soon enough when converting it to TypeScript). (thank you @sarayourfriend for helping with this step!)

To speed things up, I've prepared a `diff` with all the changes that would result from the above steps

<details>
<summary>Click to expand</summary>

```diff
diff --git a/packages/components/src/dropdown-menu/index.js b/packages/components/src/dropdown-menu/index.js
index f2ae27cd91..e347007923 100644
--- a/packages/components/src/dropdown-menu/index.js
+++ b/packages/components/src/dropdown-menu/index.js
@@ -34,22 +34,24 @@ function mergeProps( defaultProps = {}, props = {} ) {
 	return mergedProps;
 }
 
-function DropdownMenu( {
-	children,
-	className,
-	controls,
-	icon = menu,
-	label,
-	popoverProps,
-	toggleProps,
-	menuProps,
-	disableOpenOnArrowDown = false,
-	text,
-	// The following props exist for backward compatibility.
-	menuLabel,
-	position,
-	noIcons,
-} ) {
+function DropdownMenu( props ) {
+	const {
+		children,
+		className,
+		controls,
+		icon = menu,
+		label,
+		popoverProps,
+		toggleProps,
+		menuProps,
+		disableOpenOnArrowDown = false,
+		text,
+		// The following props exist for backward compatibility.
+		menuLabel,
+		position,
+		noIcons,
+	} = props;
+
 	if ( menuLabel ) {
 		deprecated( '`menuLabel` prop in `DropdownComponent`', {
 			since: '5.3',
diff --git a/packages/components/src/dropdown/index.js b/packages/components/src/dropdown/index.js
index ebee9c2b5a..09477a87e0 100644
--- a/packages/components/src/dropdown/index.js
+++ b/packages/components/src/dropdown/index.js
@@ -1,3 +1,4 @@
+// @ts-nocheck
 /**
  * External dependencies
  */
diff --git a/packages/components/src/menu-group/index.js b/packages/components/src/menu-group/index.js
index 19e182ddc7..6054bfcb3e 100644
--- a/packages/components/src/menu-group/index.js
+++ b/packages/components/src/menu-group/index.js
@@ -1,3 +1,4 @@
+// @ts-nocheck
 /**
  * External dependencies
  */
@@ -9,12 +10,8 @@ import classnames from 'classnames';
 import { Children } from '@wordpress/element';
 import { useInstanceId } from '@wordpress/compose';
 
-export function MenuGroup( {
-	children,
-	className = '',
-	label,
-	hideSeparator,
-} ) {
+export function MenuGroup( props ) {
+	const { children, className = '', label, hideSeparator } = props;
 	const instanceId = useInstanceId( MenuGroup );
 
 	if ( ! Children.count( children ) ) {
diff --git a/packages/components/src/menu-item/index.js b/packages/components/src/menu-item/index.js
index 8bed60d84d..f8a97b801b 100644
--- a/packages/components/src/menu-item/index.js
+++ b/packages/components/src/menu-item/index.js
@@ -1,3 +1,4 @@
+// @ts-nocheck
 /**
  * External dependencies
  */
@@ -16,23 +17,8 @@ import Shortcut from '../shortcut';
 import Button from '../button';
 import Icon from '../icon';
 
-/**
- * Renders a generic menu item for use inside the more menu.
- *
- * @param {Object}        props                   Component props.
- * @param {WPElement}     props.children          Element to render as child of button.
- * @param {string}        props.info              Text to use as description for button text.
- * @param {string}        props.className         Class to set on the container.
- * @param {WPIcon}        props.icon              Button's `icon` prop.
- * @param {string|Object} props.shortcut          Shortcut's `shortcut` prop.
- * @param {boolean}       props.isSelected        Whether or not the menu item is currently selected.
- * @param {string}        [props.role="menuitem"] ARIA role of the menu item.
- * @param {Object}        ref                     React Element ref.
- *
- * @return {WPComponent} The component to be rendered.
- */
-export function MenuItem(
-	{
+export function MenuItem( props, ref ) {
+	let {
 		children,
 		info,
 		className,
@@ -40,10 +26,9 @@ export function MenuItem(
 		shortcut,
 		isSelected,
 		role = 'menuitem',
-		...props
-	},
-	ref
-) {
+		...buttonProps
+	} = props;
+
 	className = classnames( 'components-menu-item__button', className );
 
 	if ( info ) {
@@ -72,7 +57,7 @@ export function MenuItem(
 			}
 			role={ role }
 			className={ className }
-			{ ...props }
+			{ ...buttonProps }
 		>
 			<span className="components-menu-item__item">{ children }</span>
 			<Shortcut
diff --git a/packages/components/src/navigable-container/container.js b/packages/components/src/navigable-container/container.js
index 68734f89be..360bcb21b9 100644
--- a/packages/components/src/navigable-container/container.js
+++ b/packages/components/src/navigable-container/container.js
@@ -1,3 +1,4 @@
+// @ts-nocheck
 /**
  * External dependencies
  */
diff --git a/packages/components/src/navigable-container/index.js b/packages/components/src/navigable-container/index.js
index e98f1b1236..b47667fd39 100644
--- a/packages/components/src/navigable-container/index.js
+++ b/packages/components/src/navigable-container/index.js
@@ -1,3 +1,4 @@
+// @ts-nocheck
 /**
  * Internal Dependencies
  */
diff --git a/packages/components/src/navigable-container/menu.js b/packages/components/src/navigable-container/menu.js
index 6e0dc794d0..df238853df 100644
--- a/packages/components/src/navigable-container/menu.js
+++ b/packages/components/src/navigable-container/menu.js
@@ -1,3 +1,4 @@
+// @ts-nocheck
 /**
  * External dependencies
  */
diff --git a/packages/components/src/navigable-container/tabbable.js b/packages/components/src/navigable-container/tabbable.js
index bc572522e9..00ba8117b8 100644
--- a/packages/components/src/navigable-container/tabbable.js
+++ b/packages/components/src/navigable-container/tabbable.js
@@ -1,3 +1,4 @@
+// @ts-nocheck
 /**
  * WordPress dependencies
  */
diff --git a/packages/components/src/tools-panel/tools-panel-header/component.tsx b/packages/components/src/tools-panel/tools-panel-header/component.tsx
index 2ef7ab290c..50f6b1564a 100644
--- a/packages/components/src/tools-panel/tools-panel-header/component.tsx
+++ b/packages/components/src/tools-panel/tools-panel-header/component.tsx
@@ -1,7 +1,3 @@
-// This can be removed once typing has been completed for DropdownMenu,
-// MenuGroup & MenuItem.
-// @ts-nocheck
-
 /**
  * External dependencies
  */
diff --git a/packages/components/tsconfig.json b/packages/components/tsconfig.json
index a0a0e879a3..cd26b356e6 100644
--- a/packages/components/tsconfig.json
+++ b/packages/components/tsconfig.json
@@ -31,6 +31,8 @@
 		"src/disabled/**/*",
 		"src/divider/**/*",
 		"src/draggable/**/*",
+		"src/dropdown/**/*",
+		"src/dropdown-menu/**/*",
 		"src/elevation/**/*",
 		"src/flex/**/*",
 		"src/flyout/**/*",
@@ -41,6 +43,9 @@
 		"src/item-group/**/*",
 		"src/input-control/**/*",
 		"src/icon/**/*",
+		"src/menu-item/**/*",
+		"src/menu-group/**/*",
+		"src/navigable-container/**/*",
 		"src/number-control/**/*",
 		"src/popover/**/*",
 		"src/range-control/**/*",
@@ -55,6 +60,7 @@
 		"src/text/**/*",
 		"src/tip/**/*",
 		"src/toggle-group-control/**/*",
+		"src/tools-panel/**/*",
 		"src/tooltip/**/*",
 		"src/truncate/**/*",
 		"src/ui/**/*",
```

</details>

Now that the issues related to dependencies should be solved, and since we've added the `src/tools-panel/**/*` folder to `tsconfig.json`, TypeScript should flag all of the remaining errors on the `ToolsPanel` set of components that still need addressing.

> Would it be best to create new PRs for MenuGroup/MenuItem etc using the doc block annotations removed from this PR as their basis? So that those can be merged quickly, or should the new PRs aim to complete a full conversion to TypeScript for those components?

I would prefer if we could achieve a full conversion of both `MenuGroup` and `MenuItem` to TypeScript in two separate PRs — both components are quite small and the conversion shouldn't be too complicated. You can definitely use the annotations that were part of this PR as a starting point! I would also recommend focusing on this PR first and opening those 2 PRs as follow-ups.

aaronrobertshaw committed:
Move ts-nocheck to dependencies and fix type errors

aaronrobertshaw commented:
Thanks for the detailed feedback, explanation and diff @ciampo! It helped me a lot.

I've applied the diff and had a go (with the assistance of @glendaviesnz) at resolving all the typing errors that were exposed. I expect that there will probably be better approaches in some cases than what I've taken. Let me know what needs improving and I'll make it happen 🙂 

As it stands now, after these updates, the ToolsPanel is functioning correctly for me in the storybook, editor and unit tests.

ciampo commented on packages/components/src/tools-panel/tools-panel-header/component.tsx:
```diff
@@ -118,7 +135,7 @@ const ToolsPanelHeader = ( props, forwardedRef ) => {
 					menuProps={ { className: dropdownMenuClassName } }
 				>
-					{ ( { onClose } ) => (
+					{ ( { onClose = () => undefined } ) => (
```
I would normally prefer not to assign a default value here, but since in this case `onClose` would come from an un-typed component (`Dropdown`), we can leave it here.

To avoid creating a new function every time, we should probably define something like `const noop = () => {};` at the very top of the file (immediately after the imports), and then use it here:

```suggestion
					{ ( { onClose = noop } ) => (
```

ciampo commented on packages/components/src/tools-panel/tools-panel-item/hook.ts:
```diff
@@ -28,14 +31,14 @@ export function useToolsPanelItem( props ) {
 	const classes = useMemo( () => {
 		return cx( styles.ToolsPanelItem, className );
-	} );
+	}, [ styles.ToolsPanelItem, className ] );
```
I don't believe that `styles.ToolsPanelItem` is required here, being an import

```suggestion
	}, [ className ] );
```

ciampo commented on packages/components/src/tools-panel/tools-panel-item/hook.ts:
```diff
@@ -28,14 +31,14 @@ export function useToolsPanelItem( props ) {
-		flagItemCustomization,
+		registerPanelItem = () => undefined,
+		deregisterPanelItem = () => undefined,
+		flagItemCustomization = () => undefined,
```
I'm not a fan of assigning empty functions here. I would instead use the optional chaining operator when invoking them later in the code — e.g. `func?.(args)`.

Alternatively, I would look into changing the `TPContext` type and make some of its values mandatory. I would then initialise the context in `context.ts` using some default no-op values for these functions.

(Hope this comment is clear enough!)

---

The same can be actually applied to the `onDeselect` and `onSelect` functions (currently defined on lines 26-27 of this file).

aaronrobertshaw replied:
> Alternatively, I would look into changing the TPContext type and make some of its values mandatory. I would then initialise the context in context.ts using some default no-op values for these functions.

✅ Done.

> The same can be actually applied to the onDeselect and onSelect functions (currently defined on lines 26-27 of this file).

This appears to be a slightly different situation to me. I've updated the defaults for this to use `noop` as suggested for other locations. Let me know if you still think it needs improving.

ciampo replied:
>  I've updated the defaults for this to use noop as suggested for other locations. Let me know if you still think it needs improving.

This looks fine to me too. For clarity, this is what I had in mind with my previous comment (which I agree was not very clear 😅 ):

```diff
diff --git a/packages/components/src/tools-panel/tools-panel-item/hook.ts b/packages/components/src/tools-panel/tools-panel-item/hook.ts
index caea9c5045..57189eea39 100644
--- a/packages/components/src/tools-panel/tools-panel-item/hook.ts
+++ b/packages/components/src/tools-panel/tools-panel-item/hook.ts
@@ -13,8 +13,6 @@ import { useContextSystem, WordPressComponentProps } from '../../ui/context';
 import { useCx } from '../../utils/hooks/use-cx';
 import type { ToolsPanelItemProps } from '../types';
 
-const noop = () => undefined;
-
 export function useToolsPanelItem(
 	props: WordPressComponentProps< ToolsPanelItemProps, 'div' >
 ) {
@@ -25,8 +23,8 @@ export function useToolsPanelItem(
 		label,
 		panelId,
 		resetAllFilter,
-		onDeselect = noop,
-		onSelect = noop,
+		onDeselect,
+		onSelect,
 		...otherProps
 	} = useContextSystem( props, 'ToolsPanelItem' );
 
@@ -95,11 +93,11 @@ export function useToolsPanelItem(
 		}
 
 		if ( isMenuItemChecked && ! isValueSet && ! wasMenuItemChecked ) {
-			onSelect();
+			onSelect?.();
 		}
 
 		if ( ! isMenuItemChecked && wasMenuItemChecked ) {
-			onDeselect();
+			onDeselect?.();
 		}
 	}, [ isMenuItemChecked, wasMenuItemChecked, isValueSet, isResetting ] );
```

I'm totally fine with either approaches, I'll leave it up to you to pick the one you think is best.

aaronrobertshaw replied:
Ah. I understand now. Thanks for the clarification. I've updated this to leverage optional chaining

ciampo commented on packages/components/src/tools-panel/tools-panel/hook.ts:
```diff
@@ -7,21 +7,32 @@ import { useEffect, useMemo, useRef, useState } from '@wordpress/element';
+	const menuItems: ToolsPanelMenuItems = { default: {}, optional: {} };
+
+	panelItems.forEach( ( panelItem: ToolsPanelItem ) => {
+		const { hasValue, isShownByDefault, label } = panelItem;
```
The changes to these lines are not necessary, since TypeScript knows that `panelItems` is of type `ToolsPanelItem[]`. The previous version of the code should still work fine:

```suggestion
	panelItems.forEach( ( { hasValue, isShownByDefault, label } ) => {
```

ciampo commented on packages/components/src/tools-panel/tools-panel/hook.ts:
```diff
@@ -81,44 +97,55 @@ export function useToolsPanel( props ) {
+	const [ menuItems, setMenuItems ] = useState< ToolsPanelMenuItems >( {
+		default: {},
+		optional: {},
+	} );
```
Is it necessary to initialise `menuItems` to this default value, instead of leaving it `undefined` ?

Also a nit: do you think it would be better, for readability's sake, to move this state declaration before `flagItemCustomization` (which calls `setMenuItems` internally) ?

aaronrobertshaw replied:
> Is it necessary to initialise menuItems to this default value, instead of leaving it undefined ?

This seemed to be the most straightforward means of addressing the typing errors reported. It initially began as an empty object. Switching to `undefined` resulted in additional errors. Later, when attempting to resolve the typing errors around dynamically updating the individual items within `menuItems` (via the `menuGroup`) we found we needed to add `default` and `optional` as mandatory properties on the `ToolsPanelMenuItems` type. This forced the initialization of the state to include those properties.

An [alternative approach](https://github.com/WordPress/gutenberg/pull/34981) has been worked up separately to this PR. For the removal of the default object when declaring the `menuItems` state, we're reducing the readability of 3 or 4 other places in the code. Personally, I like the simpler, clearer, approach in place currently.

I'm definitely happy to change that if you have a better solution though.

> Also a nit: do you think it would be better, for readability's sake, to move this state declaration before flagItemCustomization (which calls setMenuItems internally) ?

✅  Done.

ciampo replied:
Thank you for showcasing the alternative approach in a [separate PR](https://github.com/WordPress/gutenberg/pull/34981#pullrequestreview-759529019). I actually agree with you, let's keep the default value of the object as-is :)

ciampo commented on packages/components/src/tools-panel/tools-panel/hook.ts:
```diff
@@ -46,20 +57,22 @@ export function useToolsPanel( props ) {
-		const index = panelItems.findIndex( ( item ) => item.label === label );
+		const index = panelItems.findIndex(
+			( item: ToolsPanelItem ) => item.label === label
+		);
```
The changes to these lines are not necessary, since TypeScript knows that `panelItems` is of type `ToolsPanelItem[]`, and that therefore, item is of the same type. The previous version of the code should still work fine:

```suggestion
		const index = panelItems.findIndex( ( item ) => item.label === label );
```

ciampo commented on packages/components/src/tools-panel/tools-panel/hook.ts:
```diff
@@ -46,20 +57,22 @@ export function useToolsPanel( props ) {
 
 	// Allow panel items to register themselves.
-	const [ panelItems, setPanelItems ] = useState( [] );
+	const [ panelItems, setPanelItems ] = useState( <ToolsPanelItem[]>[] );
```
```suggestion
	const [ panelItems, setPanelItems ] = useState< ToolsPanelItem[] >( [] );
```

sarayourfriend commented on packages/components/src/tools-panel/tools-panel-header/component.tsx:
```diff
@@ -22,9 +18,14 @@ import MenuGroup from '../../menu-group';
+	ToolsPanelHeaderProps,
+} from '../types';
+
+const DefaultControlsGroup = ( props: ToolsPanelControlsGroupProps ) => {
```
It shouldn't be necessary to split out the destructuring of the props in this case (when they're being actually typed), that was just a strategy to prevent TS from inferring specific types for the props of the components we didn't want it to check.

aaronrobertshaw replied:
This has been addressed, thanks!

ciampo commented on packages/components/src/tools-panel/tools-panel/hook.ts:
```diff
@@ -81,44 +97,55 @@ export function useToolsPanel( props ) {
+		const filters: Array< Function > = [];
 
-		panelItems.forEach( ( item ) => {
+		panelItems.forEach( ( item: ToolsPanelItem ) => {
```
TypeScript already knows that `panelItems` is of type `ToolsPanelMenuItems`, and that therefore, `item` is of that same type

```suggestion
		panelItems.forEach( ( item ) => {
```

ciampo commented on packages/components/src/tools-panel/tools-panel/hook.ts:
```diff
@@ -81,44 +97,55 @@ export function useToolsPanel( props ) {
 	// Setup menuItems state as panel items register themselves.
 	useEffect( () => {
-		const items = generateMenuItems( { panelItems, reset: false } );
+		const items: ToolsPanelMenuItems = generateMenuItems( {
```
TypeScript already knows that `generateMenuItems` returns `ToolsPanelMenuItems`

```suggestion
		const items = generateMenuItems( {
```

ciampo commented on packages/components/src/tools-panel/tools-panel/hook.ts:
```diff
@@ -81,44 +97,55 @@ export function useToolsPanel( props ) {
 		}
 
-		const menuGroup = currentItem.isShownByDefault ? 'default' : 'optional';
+		const menuGroup: ToolsPanelMenuItemKey = currentItem.isShownByDefault
```
Since `setMenuItems` only accepts `ToolsPanelMenuItems`, typing `menuGroup` should not necessary

```suggestion
		const menuGroup = currentItem.isShownByDefault
```

ciampo commented on packages/components/src/tools-panel/tools-panel/hook.ts:
```diff
@@ -81,44 +97,55 @@ export function useToolsPanel( props ) {
+			: 'optional';
 
-		setMenuItems( {
+		const newMenuItems: ToolsPanelMenuItems = {
```
Since `setMenuItems` only accepts `ToolsPanelMenuItems`, typing `newMenuItems` should not necessary either:

```suggestion
		const newMenuItems = {
```

ciampo commented on packages/components/src/tools-panel/types.ts:
```diff
@@ -0,0 +1,69 @@
+/**
```
We should add JSDocs comments on the types describing the props for `ToolsPanel`, `ToolsPanelItem` and `ToolsPanelHeader`. The text used in the JSDocs should ideally match the documentation from the README.

Here's an example of how this is done for `ItemGroup`: https://github.com/WordPress/gutenberg/blob/85cfab17de6a23a73a6163472d1c9d85581d86c2/packages/components/src/item-group/types.ts

ciampo commented on packages/components/src/tools-panel/types.ts:
```diff
@@ -0,0 +1,69 @@
+type ResetAll = ( filters?: any[] ) => void;
+type PanelId = string;
+type ResetAllFilter = () => void;
+type Label = string;
```
This abstraction is not necessary, we can use these types directly (e.g. using `string` instead of `Label` and `PanelId`)

ciampo commented on packages/components/src/tools-panel/types.ts:
```diff
@@ -0,0 +1,69 @@
+	onSelect?: () => void;
+	className?: string;
+	resetAllFilter: ResetAllFilter;
+}
```
If `ToolsPanelItemProps` extends `ToolsPanelItem`, then it shouldn't need to specify again  `panelId` and `resetAllFilter`

ciampo commented on packages/components/src/tools-panel/types.ts:
```diff
@@ -0,0 +1,69 @@
+	panelId: PanelId;
+}
+
+export interface ToolsPanelItemProps extends ToolsPanelItem {
```
We usually use `type` instead of `interface` — e.g. 

```suggestion
export type ToolsPanelItemProps = ToolsPanelItem & {
```

@sarayourfriend do we have a strong preference here?

ciampo commented on packages/components/src/tools-panel/types.ts:
```diff
@@ -0,0 +1,69 @@
+
+export interface ToolsPanelMenuItemsConfig {
+	panelItems: ToolsPanelItem[];
+	reset: boolean;
```
Nit: it should be good if `boolean` props named suggested the props' boolean nature. Maybe we could rename it to something like `shouldReset` ?

ciampo commented on packages/components/src/tools-panel/tools-panel/hook.ts:
```diff
@@ -81,44 +97,55 @@ export function useToolsPanel( props ) {
+	const toggleItem = ( label: string ) => {
+		const currentItem: ToolsPanelItem | undefined = panelItems.find(
+			( item: ToolsPanelItem ) => item.label === label
+		);
```
The typings here should not necessary because TypeScript knows that `panelItems` is of type `ToolsPanelItem[]`, and that `Array.find()` can return such an item or undefined 

```suggestion
		const currentItem = panelItems.find( ( item ) => item.label === label );
```

ciampo commented on packages/components/src/tools-panel/tools-panel/hook.ts:
```diff
@@ -81,44 +97,55 @@ export function useToolsPanel( props ) {
 
 	const getResetAllFilters = () => {
-		const filters = [];
+		const filters: Array< Function > = [];
```
Let's not use `Function` — could this work instead?

```suggestion
		const filters: Array< () => void > = [];
```

ciampo commented on packages/components/src/tools-panel/tools-panel/hook.ts:
```diff
@@ -7,21 +7,32 @@ import { useEffect, useMemo, useRef, useState } from '@wordpress/element';
+} from '../types';
+
+const generateMenuItems = ( config: ToolsPanelMenuItemsConfig ) => {
+	const { panelItems, reset } = config;
```
You should still be able to use the previous destructuring syntax:

```suggestion
const generateMenuItems = ( { panelItems, reset }: ToolsPanelMenuItemsConfig ) => {
```

ciampo commented on packages/components/src/tools-panel/tools-panel-header/component.tsx:
```diff
@@ -11,9 +17,15 @@ import DropdownMenu from '../../dropdown-menu';
+} from '../types';
+
+const DefaultControlsGroup = ( props: ToolsPanelControlsGroupProps ) => {
+	const { items, onClose, toggleItem } = props;
```
You should be able to keep using the previous, more compact syntax here:

```suggestion
const DefaultControlsGroup = ( { items, onClose, toggleItem }: ToolsPanelControlsGroupProps ) => {
```

ciampo commented on packages/components/src/tools-panel/tools-panel-header/component.tsx:
```diff
@@ -51,7 +63,9 @@ const DefaultControlsGroup = ( { items, onClose, toggleItem } ) => {
-const OptionalControlsGroup = ( { items, onClose, toggleItem } ) => {
+const OptionalControlsGroup = ( props: ToolsPanelControlsGroupProps ) => {
+	const { items, onClose, toggleItem } = props;
+
```
You should be able to keep using the previous, more compact syntax here:

```suggestion
const OptionalControlsGroup = ( { items, onClose, toggleItem }: ToolsPanelControlsGroupProps ) => {
```

ciampo commented on packages/components/src/tools-panel/types.ts:
```diff
@@ -0,0 +1,69 @@
+	[ menuItemKey in ToolsPanelMenuItemKey ]: { [ key: string ]: boolean };
+};
+
+export interface TPContext {
```
Nit: I would call this either `Context` or `ToolsPanelContext` (to be more in line with the rest of the types)

ciampo commented on packages/components/src/tools-panel/types.ts:
```diff
@@ -0,0 +1,69 @@
+// eslint-disable-next-line no-restricted-imports
+import type { ReactNode } from 'react';
+
+type ResetAll = ( filters?: any[] ) => void;
```
Wondering if we can avoid the `any` here. What is `filters` ?

ciampo commented on packages/components/src/tools-panel/types.ts:
```diff
@@ -0,0 +1,69 @@
+	children?: ReactNode;
+	onDeselect?: () => void;
+	onSelect?: () => void;
+	className?: string;
```
`className` shouldn't be needed here, since `ToolsPanelItem` uses `WordPressComponentProps`

ciampo reviewed (commented):
Thank you @aaronrobertshaw ! I've taken a good look at the PR, now that the "setup" work is taken care ok :) Most of the comments are quite trivial

Also, adding here a comment about an unchanged line of code in `packages/components/src/tools-panel/tools-panel-item/hook.ts`:

```
	useEffect( () => {
		if ( currentPanelId === panelId ) {
			registerPanelItem( {
				hasValue,
				isShownByDefault,
				label,
				resetAllFilter,
				panelId,
			} );
		}
		return () => deregisterPanelItem( label );
	}, [ panelId ] );
```

Is it by design that `panelId` is the only dependency listed for `useEffect`?  I would have expected `currentPanelId, panelId, hasValue, isShownByDefault, label, resetAllFilter` to be all mentioned in the dependency list.

aaronrobertshaw committed:
Improve typing of ToolsPanel

Removes redundant typing, improves naming and adds JSDocs etc

aaronrobertshaw committed:
Add missing dependencies to tools panel item hook

In order to prevent excessive renders and guard against consumers not memoizing the hasValue and resetAllFilter functions, these changes include useCallback calls to handle this before adding them to the hook dependencies.

aaronrobertshaw commented:
Thank you @ciampo! I certainly appreciate the time you've taken to review this and list out the issues so clearly. 🙇 

It appears the approach I took to resolving the long list of typing errors after the setup was complete, caused a lot of redundant typing. In hindsight, cherrypicking the simplest errors first wasn't the best approach. Apologies for the extra time that cost you while reviewing. 

I believe the vast majority of the issues raised have now been addressed. 

> Also, adding here a comment about an unchanged line of code in packages/components/src/tools-panel/tools-panel-item/hook.ts:
> 
> Is it by design that panelId is the only dependency listed for useEffect? I would have expected currentPanelId, panelId, hasValue, isShownByDefault, label, resetAllFilter to be all mentioned in the dependency list.

Initially, this hook was only to run once. An edge case was found while copy and pasting blocks that used SlotFills to inject panel items. As a result, the `panelId` was added to the dependencies. I've added the remaining dependencies now. 

It's worth noting though that to guard against the possibility of consumers not memoizing the `hasValue` and `resetAllFilter` functions, I've wrapped them in a `useCallback`. Without this, the panel's display of items breaks due to the panel items registering/deregistering excessively. Let me know if you see any issues with this approach.

Thanks again for all your time and effort 👍

ciampo commented on packages/components/src/tools-panel/types.ts:
```diff
@@ -0,0 +1,130 @@
+export type ToolsPanelContext = {
+	panelId?: string;
+	menuItems?: ToolsPanelMenuItems;
+	hasMenuItems?: number;
```
I would expect a variable called `hasMenuItems` to be a `boolean` — I can see 3 options here to improve this:

1. rename the variable (e.g. `howManyMenuItems` or any better name that you can think of)
2. remove the variable (and use `menuItems.length` directly in the code)
3. keep the variable name, but change it to be `boolean` (and change its computation to something like `panelItems.length > 0`)

aaronrobertshaw replied:
I've enforced `hasMenuItems` to be a boolean. Thanks.

ciampo commented on packages/components/src/tools-panel/context.ts:
```diff
@@ -0,0 +1,20 @@
+	registerPanelItem: noop,
+	deregisterPanelItem: noop,
+	flagItemCustomization: noop,
+} );
```
Have you considered marking more of the `ToolsPanelContextType` props as non-optional, and giving them a default value here? For example:
- `menuItems` could be initialised to the to that same default value with empty objects used in other parts of the code
- `hasMenuItems` could be initialised to `0` (or `false`, pending another review comment)
- `isResetting` could be initialised to `false` (I guess?)

`panelId` is the only variable that could be left optional, and therefore initially `undefined`.


Also, this comment is more of a personal preference, so feel free to ignore it :)

aaronrobertshaw replied:
All the ToolsPanelContext props should be mandatory now except `panelId`. Given they should always be there, it makes sense to me. Thanks again.

ciampo reviewed (commented):
> Thank you @ciampo! I certainly appreciate the time you've taken to review this and list out the issues so clearly. 🙇

And I appreciate your patience and determination in addressing all feedback so quickly. It's great to see you and Glen approaching this TypeScript refactor with little previous experience and still managing to produce great work!

> Initially, this hook was only to run once. An edge case was found while copy and pasting blocks that used SlotFills to inject panel items. As a result, the panelId was added to the dependencies. I've added the remaining dependencies now.
> It's worth noting though that to guard against the possibility of consumers not memoizing the hasValue and resetAllFilter functions, I've wrapped them in a useCallback. Without this, the panel's display of items breaks due to the panel items registering/deregistering excessively. Let me know if you see any issues with this approach.

Thank you for the background — I've tested the component in Storybook and didn't notice any weird behaviour.

I've left a couple more comments, but I don't consider them blocking. As the last step, I think it'd be good to update the README with some more specific TypeScript types (e.g. `resetAll` could be documented as of type `() => void` [instead of `function`](https://github.com/WordPress/gutenberg/blob/3b14a920de2957868954f7ffd018cf0ff1845d48/packages/components/src/tools-panel/tools-panel/README.md?plain=1#L85)).

I've also noticed a little error when [defining `panelId` as a `function`](https://github.com/WordPress/gutenberg/blob/3b14a920de2957868954f7ffd018cf0ff1845d48/packages/components/src/tools-panel/tools-panel/README.md?plain=1#L77) instead of a `string`.

Finally, I've noticed that, when resetting values in Storybook, the "unit" of the Unit Controls does not get reset to "px" — is that by design, or should it be addressed (in a separate PR / issue)?

aaronrobertshaw committed:
Use optional chaining for onSelect and onDeselect calls

aaronrobertshaw committed:
Add TypeScript types for function props in READMEs

aaronrobertshaw committed:
Enforce boolean for hasMenuItems in ToolsPanelContext

aaronrobertshaw committed:
Make all properties except panelId mandatory in ToolsPanelContext

aaronrobertshaw commented:
I've made the remaining updates to this. It should be getting close now I think. Thanks yet again @ciampo  🙂 

> Finally, I've noticed that, when resetting values in Storybook, the "unit" of the Unit Controls does not get reset to "px" — is that by design, or should it be addressed (in a separate PR / issue)?

I believe this is the behaviour of the `UnitControl`. I'd need to check but from memory, the `UnitControl` will store the currently selected unit in internal state. Clearing the value doesn't change this. I'm assuming because the user can select a unit before entering any value at all.

Given this, I think any changes there should be addressed in a separate issue/PR, especially given the widespread use of UnitControl.

ciampo reviewed (approved):
This is looking good, excellent job everyone involved! Let's 🚀 !

(just noticing the CI errors, maybe a rebase would fix them?)

> I believe this is the behaviour of the UnitControl. [...] Given this, I think any changes there should be addressed in a separate issue/PR, especially given the widespread use of UnitControl.

Yeah, 100% — Let me know if you need any help with that.
